mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java

```diff
@@ -0,0 +1,79 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.search;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+
+public interface ValueMatcher<T> {
+
+    boolean matches(T referenceValue);
+
+    interface SingleValueMatcher<T> extends ValueMatcher<T> {
+        T testedValue();
+    }
+
+    interface ListMatcher<T> extends ValueMatcher<List<T>> {
+        T testedValue();
+    }
+
+    interface Equals<T> extends SingleValueMatcher<T> {
+        @Override
+        default boolean matches(T referenceValue) {
+            return referenceValue.equals(testedValue());
+        }
+    }
+
+    interface StringContains extends SingleValueMatcher<String> {
```

Comment:
I don't agree about your use of interfaces, it makes things more complex than needed.
If you want to pass a function to a class, just do that but don't overuse interface inheritance.

Reply:
I removed multiple interfaces by classes and static methods.